quodlibet/quodlibet/ext/events/waveformseekbar.py

```diff
@@ -0,0 +1,277 @@
+        self._rms_vals = []
+
+    def _on_bus_message(self, bus, message):
+        if message.type == Gst.MessageType.ERROR:
+            error, debug = message.parse_error()
+            print("Error received from element {name}: {error}".format(
+                name=message.src.get_name(), error=error))
+            print("Debugging information: {}".format(debug))
+        elif message.type == Gst.MessageType.ELEMENT:
+            structure = message.get_structure()
+            if(structure.get_name() == "level"):
+                rms_dB = structure.get_value("rms")
+                # Calculate average of all channels (usually 2)
+                rms_dB_avg = sum(rms_dB) / len(rms_dB)
+                # Normalize dB value to value between 0 and 1
+                rms = pow(10, (rms_dB_avg / 20))
+                self._rms_vals.append(rms)
+            else:
+                # Shouldn't happen
+                print("Not Level")
+        elif message.type == Gst.MessageType.EOS:
+            self._pipeline.set_state(Gst.State.NULL)
+            self._waveform_scale.update(self._rms_vals, self._player)
+
+    def _on_destroy(self, *args):
+        self._tracker.destroy()
+
+    def _on_song_changed(self, library, songs, player):
+        self._update(player)
+        if player.info:
+            self._create_waveform(player.info("~filename"))
+
+    def _on_tick(self, tracker, player):
+        self._update(player)
+
+    def _on_seekable_changed(self, player, *args):
+        self._update(player)
+
+    def _on_player_seek(self, player, song, ms):
+        self._update(player)
+
+    def _on_song_started(self, player, song):
+        self._update(player)
+
+    def _update(self, player):
+        if player.info:
+            # Position in ms, length in seconds
+            position = player.get_position() / 1000
+            length = (player.info("~#length"))
+            remaining = length - position
+
+            if length == 0:
+                return
+
+            self._waveform_scale.set_position(position / length)
+            self._elapsed_label.set_time(position)
+            self._remaining_label.set_time(remaining)
+            self._remaining_label.set_disabled(not player.seekable)
+            self._elapsed_label.set_disabled(not player.seekable)
+
+            self.set_sensitive(player.seekable)
+
+
+class WaveformScale(Gtk.Widget):
+    """ The waveform widget. """
+
+    # __gtype_name__ = 'WaveformScale'
+
+    _rms_vals = []
+
+    def __init__(self, *args, **kwds):
+        super(WaveformScale, self).__init__(*args, **kwds)
+        self.set_size_request(40, 40)
+        self.position = 0
+
+    def update(self, rms_vals, player):
+        self._rms_vals = rms_vals
+        self.queue_draw()
+        self._player = player
+
+    def do_draw(self, cr):
+        # Paint the background
+        context = self.get_style_context()
+        bg_color = context.get_background_color(Gtk.StateFlags.NORMAL)
```

Comment:
This no longer works on newer gtk ( and is probably the source of 100% CPU) You need to do:

```
context.save()
context.set_state(Gtk.StateFlags.NORMAL)
bg_color = context.get_background_color(context.get_state())
context.restore()
```